### PR TITLE
Initial update for patch 5.1

### PR DESCRIPTION
**Pull request by ryan-dyer-sp**
I took the liberty of making an initial swipe at updating for patch 5.1.  

There were certain things I wasnt sure how to handle.  ie. the new traits which boost basic and rapid synth.  So I went with the approach of treating them as a different skill but with a display name of the lower level version.  

I also did not obtain any new images or modify any.

This is more intended to be something I and others can use until doxxx get's back from vacation and does things better. @doxxx feel free to use this as a starting spot when you get back if you like.

If anyone finds issues (via cloning and running locally yourself), please feel free to post a comment to the PR I will try and address.

### CODE REVIEW

**Comment by tanpro260196** [comment]
Can't solve for anything. Stuck like this:
image

I tried to clear cookies and switch browser. Doesn't help.

**Comment by strevel** [comment]
chrome_2019-11-02_16-43-26
setting level in crafter attribute to anything higher than 70 makes it unusable
works okay when 70 or below

**Comment by ryan-dyer-sp** [comment]
Please hit F12 and copy the console output error and post it.

**Comment by tanpro260196** [comment]
> worker.js:36 TypeError: Cannot read property 'shortName' of undefined
    at ApplySpecialActionEffects (ffxivcraftmodel.js:631)
    at UpdateState (ffxivcraftmodel.js:776)
    at MonteCarloStep (ffxivcraftmodel.js:1067)
    at MonteCarloSequence (ffxivcraftmodel.js:1254)
    at start (worker.js:191)
    at self.onmessage (worker.js:21)
self.onmessage @ worker.js:36
worker.js:40 Uncaught TypeError: Cannot read property 'logOutput' of undefined
    at self.onmessage (worker.js:40)

I got this error when I was trying to solve for my lv70 weaver.

**Comment by ryan-dyer-sp** [comment]
More changes made.  Thanks for testing and if you see other issues, please post your crafters stats and whether or not its a specialist as well as what you're attempting to craft.  In the case above, it was due to having a specialist.  As my highest crafter is 53, I don't have specialization and didnt know what it provides (+15 cp) until today.

**Comment by tanpro260196** [comment]
image
Another problem, the recipes durability is 70. Master's Mend should be use after 6 action, not 7 as it will fail if durability reach 0.

> /ac "Reflect" <wait.3>
/ac "Basic Touch" <wait.3>
/ac "Basic Touch" <wait.3>
/ac "Basic Touch" <wait.3>
/ac "Basic Touch" <wait.3>
/ac "Basic Touch" <wait.3>
/ac "Basic Touch" <wait.3>
/ac "Master's Mend" <wait.3>
/ac "Basic Touch" <wait.3>
/ac "Basic Touch" <wait.3>
/ac "Basic Touch" <wait.3>
/ac "Master's Mend" <wait.3>
/ac "Basic Touch" <wait.3>
/ac "Basic Touch" <wait.3>
/echo Macro #1 complete <se.1>

**Comment by werlej** [comment]
first of all, thanks for putting up a temp site for this. 
some things I noticed:
- looks like its using the old CP cost for Rapid Synth for later levels
- Byregot's Blessing is missing 
- Final Appraisal is missing
- if specialist is checked it still gives the +15 CP (which is no longer a thing)

**Comment by doxxx** [comment]
Awesome! Thanks for doing this!

**Comment by ryan-dyer-sp** [comment]
> first of all, thanks for putting up a temp site for this.
> some things I noticed:
> 
> * looks like its using the old CP cost for Rapid Synth for later levels
> * Byregot's Blessing is missing
> * Final Appraisal is missing
> * if specialist is checked it still gives the +15 CP (which is no longer a thing)

According to the patch notes:
Soul of the Crafter now also increases CP by 15. is this something different than being a specialist?

I'm leaving doxxx to deal with final appraisal and heuristic  related changes.  He understands it alot better than the 2 hours I've spent looking at this code.  Even the way I've made changes to muscle memory and reflect likely arent entirely proper and may be impacting how the final results come out.  But its all better than the nothing that we have currently ;p

blessing and rapid synth fix.

**Comment by AricStewart** [comment]
Hello I am level 80 in all crafting areas and am happy to help and test. Tell me what you want tests and I can give you results.
One thing of note is that the progress and quality numbers on the simulator window are wrong.
My crafter (craft 2175, control 2009) says in the crafting pane for a level 80 item  (dwarven mythril hatchet) progress 353 and quality 377. the simulator says progress 264 and quality 277.

for my specialist (same gear) (craft 2195, control 2029)  Dwarven mythril ring has progress 356 and quality 381 in game,  simulator says 356, 381

**Comment by AricStewart** [comment]
Apology for the previously incorrect message. I had imported my stats to the test site wrong. 

However I am a level 80 crafter happy to help with testing! Tell me how I can can be helpful.

**Comment by werlej** [comment]
the thing with soul of the crafter is that its always there so its just applied to your base stats, which is likely the value people will put into the tool (i.e. if my CP for a non-specialist is 460 in the character menu, it will be 475 on a specialist in the character menu)

**Comment by ryan-dyer-sp** [comment]
i see.  so does it serve any purpose to have the option to select that you have a craft job specialized then?

**Comment by werlej** [comment]
in my opinion it doesn't since the only specialist action you have now is 'Careful Observation' (but costs a crafter's delineation to use and not entirely sure how different it is from 'Observe' aside from the 0 CP cost)

**Comment by tanpro260196** [comment]
> /ac "Reflect" <wait.3>
/ac "Ingenuity" <wait.2>
/ac "Delicate Synthesis" <wait.3>
/ac "Prudent Touch" <wait.3>
/ac "Prudent Touch" <wait.3>
/ac "Ingenuity" <wait.2>
/ac "Prudent Touch" <wait.3>
/ac "Prudent Touch" <wait.3>
/ac "Delicate Synthesis" <wait.3>
/ac "Prudent Touch" <wait.3>
/ac "Prudent Touch" <wait.3>
/ac "Ingenuity" <wait.2>
/ac "Prudent Touch" <wait.3>
/ac "Byregot's Blessing" <wait.3>
/echo Macro #1 complete <se.1>

Ingenuity count down is 5. And it will fail if use too close to each other like this (you have to wait the the buff to expire before refresh it.

**Comment by EpicElric** [comment]
The new Custom Deliveries recipes for 5.1 are missing.

**Comment by Kaiay** [comment]
no skill call "倹約II" is inside the game.
it should be call "長期倹約" LV:47 CP:98
And also the icon of this skill is wrong.(I know it is just a test version sorry)
Also found that not all the skill is translated in Optimizer so try to provide that skill's Japanese name.
Trained Eye -> 匠の早業 LV:80 CP:250
Reflect -> 真価 LV:69 CP:24

**Comment by ryan-dyer-sp** [comment]
> > /ac "Reflect" <wait.3>
> > /ac "Ingenuity" <wait.2>
> > /ac "Delicate Synthesis" <wait.3>
> > /ac "Prudent Touch" <wait.3>
> > /ac "Prudent Touch" <wait.3>
> > /ac "Ingenuity" <wait.2>
> > /ac "Prudent Touch" <wait.3>
> > /ac "Prudent Touch" <wait.3>
> > /ac "Delicate Synthesis" <wait.3>
> > /ac "Prudent Touch" <wait.3>
> > /ac "Prudent Touch" <wait.3>
> > /ac "Ingenuity" <wait.2>
> > /ac "Prudent Touch" <wait.3>
> > /ac "Byregot's Blessing" <wait.3>
> > /echo Macro #1 complete <se.1>
> 
> Ingenuity count down is 5. And it will fail if use too close to each other like this (you have to wait the the buff to expire before refresh it).

I don't believe I've done anything in my changes which would warrant this behavior.  ie. i believe it likely that this behavior already existed.

**Comment by ryan-dyer-sp** [comment]
> no skill call "倹約II" is inside the game.
> it should be call "長期倹約" LV:47 CP:98
> And also the icon of this skill is wrong.(I know it is just a test version sorry)
> Also found that not all the skill is translated in Optimizer so try to provide that skill's Japanese name.
> Trained Eye -> 匠の早業 LV:80 CP:250
> Reflect -> 真価 LV:69 CP:24

Sorry but as I dont read or write Japanese, I'm not going to make translation changes as part of my PR.  Please feel free to write this up as a different bug/issue or submit your own PR.

**Comment by ryan-dyer-sp** [comment]
> in my opinion it doesn't since the only specialist action you have now is 'Careful Observation' (but costs a crafter's delineation to use and not entirely sure how different it is from 'Observe' aside from the 0 CP cost)

I have removed the 15 cp assuming people will inherently put the CP into the tool including this value.

**Comment by Sephiroth018** [comment]
Ingenuity's cost is wrong, it has been lowered to 22 CP

**Comment by Coartana** [comment]
When importing a settings file that included the Cross-Class Action Macro setting enabled (from the master branch, not this pull request), the section for that macro still exists on the solver page, but cannot be unset. Would recommend clearing the flag as part of import or removing the section for the macro in the solver tab.

**Comment by doxxx** [comment]
Sorry for the wait, I came down with the flu on vacation :P and took a long time to recover. How is this coming?

**Comment by tanpro260196** [comment]
> Sorry for the wait, I came down with the flu on vacation :P and took a long time to recover. How is this coming?

It's functional so far. Not sure how good it is on latest contents since I didn't participate in those. But on 80* or lower it seem to works fine.

**Comment by ryan-dyer-sp** [comment]
I've addressed most functional issues with the 5.1 change; at least those that have been reported.  I didn't want to combine any item changes into it as I felt the PR as probably large enough.  I also sent you a message on linkedin, not sure if you saw it.  This also doesnt contain any icon changes.  I tried using the item generator repo but the output didnt land in the same format as what is used by this tool, so not sure if i did something wrong or perhaps it serves a different purpose.

File: app/js/actions.js [hunk 1]
```diff
@@ -30,78 +30,78 @@ var AllActions = {
...
-    standardSynthesis: new Action(  'standardSynthesis',    'Standard Synthesis',   10,     15,  0.9, 0.0, 1.5, 'immediate',   1,  'All',          31),
-    carefulSynthesis: new Action(   'carefulSynthesis',     'Careful Synthesis',    10,      0,  1.0, 0.0, 0.9, 'immediate',   1,  'Weaver',       15),
-    carefulSynthesis2: new Action(  'carefulSynthesis2',    'Careful Synthesis II', 10,      0,  1.0, 0.0, 1.2, 'immediate',   1,  'Weaver',       50),
-    rapidSynthesis: new Action(     'rapidSynthesis',       'Rapid Synthesis',      10,      0,  0.5, 0.0, 2.5, 'immediate',   1,  'Armorer',      15),
-    flawlessSynthesis: new Action(  'flawlessSynthesis',    'Flawless Synthesis',   10,     15,  0.9, 0.0, 1.0, 'immediate',   1,  'Goldsmith',    37),
-    pieceByPiece: new Action(       'pieceByPiece',         'Piece by Piece',       10,     15,  0.9, 0.0, 1.0, 'immediate',   1,  'Armorer',      50),
+    standardSynthesis: new Action(  'standardSynthesis',    'Basic Synthesis',      10,      0,  1.0, 0.0, 1.2, 'immediate',   1,  'All',          31),
```
Comment (doxxx) [inline app/js/actions.js]: Is there a reason the name was changed to `Basic Synthesis`?

Comment (ryan-dyer-sp) [reply app/js/actions.js]: They got rid of levels of ability for this.  You now unlock trait at some level which upgrades your existing abilities.  Riding in vehicle or I'd post the trait name and description of you cared.

Comment (ryan-dyer-sp) [reply app/js/actions.js]: From a code perspective there should likely be something in the heuristics which prevents it from attempting basic synth when standard is available.  There may already be but once again on the road so can't easily investigate

Comment (doxxx) [reply app/js/actions.js]: Looks like the same has happened to Rapid Synthesis.

(see https://www.xivcraftingguide.com/51-patch-changes)

**Comment by doxxx** [comment]
@ryan-dyer-sp I've removed all the commented out code and fixed a few other things. Please have a look and let me know if you're okay with these changes, then I'll merge the PR. After that, further fixes and tweaks can be made in new PRs of hopefully smaller scope. :)

File: app/js/actions.js (hunk 1 above)
Comment (doxxx) [reply app/js/actions.js]: I added the level to the action tooltip to make it easier to distinguish these actions for now, until a better solution can be determined.

File: app/js/ffxivcraftmodel.js [hunk 2]
```diff
@@ -1664,30 +1370,19 @@ function heuristicSequenceBuilder(synth) {
...
 
-    // if ((effCrafterLevel < effRecipeLevel) && tryAction('ingenuity2')) {
-    //     pushAction(subSeq1, 'ingenuity2');
-    //     if (Ing2RecipeLevelTable[effRecipeLevel]) {
-    //         //effRecipeLevel = Ing2RecipeLevelTable[effRecipeLevel];
-    //     }
-    // }
     if ((effCrafterLevel < effRecipeLevel) && tryAction('ingenuity')) {
         pushAction(subSeq1, 'ingenuity');
         if (Ing1RecipeLevelTable[synth.recipe.level]) {
```
Comment (ryan-dyer-sp) [inline app/js/ffxivcraftmodel.js]: Given this has a noop expression, should the conditional just be removed?

**Comment by ryan-dyer-sp** [comment]
one minor comment, otherwise it all looks fine to me.

**Comment by doxxx** [comment]
Thank you for all your hard work!

**Comment by doxxx** [comment]
Uploaded to beta site.

**Comment by ryan-dyer-sp** [comment]
beta site?

**Comment by doxxx** [comment]
ffxiv-beta.lokyst.net

**Pull request merged**